### PR TITLE
x for grids

### PR DESCRIPTION
**Pull request by bkrmendy**
https://github.com/user-attachments/assets/7627f89e-c87b-4246-9ac5-1f70ca4e254d

## Problem
The x shortcut isn't aware of grids

## Fix
Special case grid elements in the x shortcut.

### How it works
If an element is a grid child and it's absolute positioned (for example after a reparent or drawing), pressing x removes the `position: absolute`, absolute positioning prop, and any explicit width/height props on the element, and sets `gridRow`/`gridColumn`/`gridRowStart`/gridRowEnd`/`gridColumnStart`/gridColumnEnd` in way that snaps the element into the grid cells it occupies.

When an element is a non-position-absolute grid child, and x is pressed, `position: absolute` is set, the element is sized to its visual dimensions with `width`/`height`, and `top: 0` / `left: 0` is added. Making it possible to absolute resize the element after this is left to a follow-up PR. 

### Details
- Due to a circular dependency, a lot of grid cell related code was factored out into `grid-cell-utils.ts`

### CODE REVIEW

**Review by ruggi — APPROVED** [review]

File: editor/src/components/inspector/inspector-common.ts [hunk 1]
```diff
@@ -1401,3 +1412,122 @@ export function getConstraintsIncludingImplicitForElement(
...
+):
+  | null
+  | { type: 'contained'; gridRow: number; gridColumn: number }
+  | {
+      type: 'span'
+      gridRowStart: number
+      gridRowEnd: number
+      gridColumnStart: number
+      gridColumnEnd: number
+    } {
```
Comment (Rheeseyb) [inline editor/src/components/inspector/inspector-common.ts:1428]: Please don't do this in the return type. We should just create a type for each of these

File: editor/src/components/inspector/inspector-common.ts [hunk 2]
```diff
@@ -1401,3 +1412,122 @@ export function getConstraintsIncludingImplicitForElement(
...
+  if (instance == null) {
+    return null
+  }
+
+  const cellBounds = getGridElementBounds(instance, canvasContext)
+  if (cellBounds == null) {
+    return null
+  }
+
+  if (MetadataUtils.isPositionAbsolute(instance)) {
```
[the remark this answers is not in the document]
Comment (bkrmendy) [reply editor/src/components/inspector/inspector-common.ts:1488]: https://github.com/concrete-utopia/utopia/pull/6332/commits/a435f1f37f9a9a92d835f23a110400fb5b995d6d

File: editor/src/components/inspector/inspector-common.ts (hunk 1 above)
Comment (bkrmendy) [reply editor/src/components/inspector/inspector-common.ts]: https://github.com/concrete-utopia/utopia/pull/6332/commits/78bed2c3dd550ae348c6d2a91195166a8080b6f1

File: editor/src/components/editor/shortcuts.spec.browser2.tsx [hunk 3]
```diff
@@ -259,6 +259,187 @@ describe('shortcuts', () => {
...
+        data-uid='child'
+        data-testid='child'
+          style={{
+            backgroundColor: '#0162ff',
+            position: 'absolute',
+            left: 13,
+            top: 19,
+            width: 74,
+            height: 59,
+            gridColumn: 3,
```
Comment (balazsbajorics) [inline editor/src/components/editor/shortcuts.spec.browser2.tsx:309]: please add a test that converts an absolute child withot gridColumn and gridRow

**Review by balazsbajorics — APPROVED** [review]

File: editor/src/components/editor/shortcuts.spec.browser2.tsx (hunk 3 above)
Comment (bkrmendy) [reply editor/src/components/editor/shortcuts.spec.browser2.tsx:309]: https://github.com/concrete-utopia/utopia/pull/6332/commits/4c5f050ed4e36e790fefeb99e44bc614954f0a16

**Pull request merged**